sematic/log_reader.py

```diff
@@ -422,40 +505,78 @@ def line_stream_from_log_directory(
         key=lambda path_key: int(
             path_key.replace(directory, "").replace(".log", "".replace("/", ""))
         ),
+        reverse=reverse,
     )
-    found_cursor_file = cursor_file is None
-    found_cursor_line = cursor_line_index is None
-    for log_file in log_files:
-        if log_file == cursor_file:
-            found_cursor_file = True
-        if not found_cursor_file:
-            continue
+    if len(log_files) == 0:
+        yield from []
+        return
+    index_of_log_file = 0
+    if cursor_file is not None:
+        if cursor_file not in log_files:
+            raise ValueError(f"Cursor referenced invalid log file {cursor_file}.")
+        index_of_log_file = log_files.index(cursor_file)
+
+    cursor_file = log_files[index_of_log_file]
+    log_files_to_traverse = log_files[index_of_log_file:]
+    for log_file in log_files_to_traverse:
+        start_line_index = DEFAULT_END_INDEX if reverse else DEFAULT_START_INDEX
+        if log_file == cursor_file and cursor_line_index is not None:
+            start_line_index = cursor_line_index
+
         text_stream: Iterable[str] = storage.get_line_stream(log_file)
+        yield from _stream_from_text_stream_from_index(
+            log_file, start_line_index, reverse, text_stream
+        )
+
+
+def _stream_from_text_stream_from_index(
+    log_file: str, start_line_index: int, reverse: bool, text_stream: Iterable[str]
+) -> Iterable[LogLine]:
+    """Starting at the given index within the file, stream logs in desired direction.
+
+    When going forward, the line pointed to by the index will be included. When going in
+    the reverse direction, the line pointed to by the index will be omitted.
+    """
+    # For purposes of this function, "preceding" refers to indices
+    # lower than the start index, regardless of traversal direction.
+    lines_preceding_cursor: List[LogLine] = []
+    if reverse:
         for i_line, line in enumerate(text_stream):
-            if (
-                (not found_cursor_line)
-                and cursor_line_index is not None
-                and i_line < cursor_line_index
-            ):
-                continue
-            found_cursor_line = True
-            yield LogLine(
+            as_log_line = LogLine(
+                source_file=log_file,
+                source_file_index=i_line,
```

Comment:
We talked about including a **line key** alongside the line text, is this `source_file_index` going to contribute to that task?

Can you explain how it would be done? Is it done in this PR or separate? I wonder if both `source_file` + `source_file_index` will participate in forming the line key?

Reply:
I will do it in a separate PR, and it will be derived from the source file and index